[POST] Changer le type d'une clef primaire non référencée à BIG INTEGER (#19)
1024pix/pix-engineering: _posts/2021-08-31-modifier-type-clef-primaire-non-referencee-vers-biginit.md

---
layout: post
title:  "Passer une PK en BIG INTEGER"
date:   2021-08-30 10:10:42 +0200
categories: bdd
excerpt: "Changer le type d'une clef primaire de INTEGER à BIG INTEGER en minimisant l'indisponibilité"
authors: pierre_top
---

## TL;DR

Notre table la plus volumineuse aurait atteint son remplissage maximal dans quelques mois à cause du type de la clef primaire. Pour éviter cela, nous souhaitions changer de type en minimisant l'impact utilisateur, en visant des migrations de BDD "zero-downtime".

Nous avons réussi (pour les impatients, la solution est [ici](implementation)) et partageons avec vous nos apprentissages.

Ceci dit, cela a eu lieu dans des circonstances particulières :

- table non référencée par d'autres tables;
- traffic réduit (vacances).

> Ce chantier continuera sur un autre PK, sans ces deux circonstances, et nous partagerons également les résultats dans un autre post.

Reviewer: ```suggestion
Ce chantier continuera sur une autre PK, sans ces deux circonstances, et nous en partagerons également les résultats dans un autre post.
```

Pour valider les contraintes en amont :

- ajouter une contrainte NOT NULL dès la création de la colonne ;
- créer un index UNIQUE de manière concurrente (`CREATE INDEX CONCURRENTLY`)[https://www.postgresql.org/docs/current/sql-createindex.html].

Lors de la création de la PK, il suffit de préciser l'index UNIQUE (PostgreSQL détecte la contrainte NOT NULL).
`ALTER TABLE "foo" ADD CONSTRAINT "foo_pkey" PRIMARY KEY USING INDEX "foo_id_unique`

## Implémentation

La solution est présentée en pseudo SQL/NodeJS sur une table imaginaire, avec les liens vers les commits réels.

On part de là

``` sql
CREATE TABLE foo (ID INTEGER PRIMARY KEY);
INSERT INTO foo generate_series(1, 1e6);
```

Voilà la solution native, avec accès exclusif sur la table pendant 8 heures.

```sql
ALTER TABLE foo COLUMN idBigInteger BIG INTEGER;
```

Tout commence avec une migration lancée lors du déploiement de release
<https://github.com/1024pix/pix/pull/3357/files#diff-1eecfd94150cca766655d3786c7dc12e3b6bc8fa60be3a0c0740424ab02c157a>

```sql
ALTER TABLE foo ADD COLUMN idBigInteger NOT NULL DEFAULT -1;
CREATE TRIGGER WHEN INSERT INTO foo FOR EACH ROW :NEW.bigIntId = :new.id;
```

> Ensuite, le script de migration inclut dans la PR est lancé manuellement (durée: 48h).

Reviewer: ```suggestion
Ensuite, le script de migration inclus dans la PR est lancé manuellement (durée: 48h).
```

<https://github.com/1024pix/pix/pull/3357/files#diff-d89cb20637cb9b836bce4902b46db958e0209c2a0de96e7dc81a0c43015680b4>

```javascript
const maxid = await client.query('SELECT MAX(id) INTO maxId FROM foo');
for (const i = 0; i <= maxId; i++){
  await client.query(`UPDATE foo SET idBigInteger = id WHERE id BETWEEN i AND (i + ${chunkSize} -1)`);
}
await client.query('CREATE UNIQUE INDEX CONCURRENTLY indexidBigInteger');
```

> Pour finir, une dernière migration lancée lors du déploiement de release

Reviewer: ```suggestion
Pour finir, une dernière migration est lancée lors du déploiement de la release
```

<https://github.com/1024pix/pix/pull/3364/files#diff-9985cafe7c9915c64927036180071c804f75aad26a66686efd1758d242b819d1>

```sql
BEGIN TRANSACTION;
LOCK foo WITH ACCESS EXCLUSIVE;
ALTER SEQUENCE seqPkFoo TYPE BIG INTEGER;
ALTER TABLE foo ALTER COLUM bigIntId SET DEFAULT seqPkFoo.nextval();
ALTER TABLE foo DROP COLUMN id;
ALTER TABLE foo RENAME COLUMN idBigInteger TO id;
ALTER TABLE foo ADD PRIMARY KEY id USING indexidBigInteger;
COMMIT TRANSACTION;
```

## Exécution en production et perspectives

> ### Le plan se déroule sans accrocs

Reviewer: ```suggestion
### Le plan se déroule sans accroc
```

La migration préalable s'est déroulée sans erreur.

La migration des données proprement dite a pris environ 24h.

L'exposition de la nouvelle colonne, lors du déploiement, a pris un temps négligeable pour les utilisateurs finaux (< 1 seconde).

```shell
2021-08-27 14:10:40.368443479 +0200 CEST [postdeploy-2719] > pix-api@3.92.0 postdeploy /app
2021-08-27 14:10:40.368393446 +0200 CEST [postdeploy-2719]
2021-08-27 14:10:41.566685265 +0200 CEST [postdeploy-2719] Batch 188 run: 1 migrations
2021-08-27 14:10:42.176473412 +0200 CEST [manager] container [postdeploy-2719] (6128d63ffcff1013ac9e57c1) has stopped
```

### Mais

#### Un impact non négligeable du traitement de migration sur le temps de réponse

Bien que l'opération soit un succès, nous avons couru plusieurs risques :

- la migration des données (phase `UDPATE`) a sensiblement ralenti le temps de réponse global ;

> - la migration des données (phase `CREATE INDEX CONCURENTLY`) a acquis un verrou exclusif sur une période de x minutes.

Reviewer: x minutes ?

Author: Oui, c'est à compléter avec les chiffres (@jonathanperret les a mais pas moi)

Reviewer: Ok j'avais peur que ce soit un oubli.

Reviewer 2: Je n'ai pas voulu parler de verrou exclusif (à ma connaissance il n'y en a pas eu, ou alors de l'ordre de quelques millisecondes) mais d'une ou plusieurs phases où l'impact sur les ressources était tel que les temps de réponse de l'application étaient affectés au-delà du supportable.

Author: :fearful: Bon, on va pouvoir regarder ça ensemble histoire que je ne raconte pas de bêtises..

Author: J'ai corrigé, en prenant le temps de réponse moyen du routeur API: temps de réponse d'une seconde pendant une heure et demie pendant le CREATE INDEX.
A cette occasion, j'ai vu qu'il y avait aussi eu 4 autres pics de 1s, qui ont duré 30 minutes, pendant l'UPDATE.

---
layout: post

> title:  "Passer une PK de INT en BIGINIT"

Reviewer 3: ```suggestion
title:  "Passer une PK de INT en BIGINT"
```

Reviewer 3: on peut ajouter dans le titre que ça concerne une base PG ou bien c'est pas nécessaire tu penses?

Author: C'est bien de préciser je pense => Corrigé

   - la capacité de reproduire un trafic similaire à la production ;
   - le fait que l'environnement soit sur une plateforme dédiée, avec une disponibilité des ressources partagée (ex: I/O) non déterministe.

## Exécution en production et perspectives

### Le plan se déroule sans accroc

La migration préalable s'est déroulée sans erreur.

La migration des données proprement dite a pris environ 24h.

L'exposition de la nouvelle colonne, lors du déploiement, a pris un temps négligeable pour les utilisateurs finaux (< 1 seconde).

```shell
2021-08-27 14:10:40.368443479 +0200 CEST [postdeploy-2719] > pix-api@3.92.0 postdeploy /app
2021-08-27 14:10:40.368393446 +0200 CEST [postdeploy-2719]
2021-08-27 14:10:41.566685265 +0200 CEST [postdeploy-2719] Batch 188 run: 1 migrations
2021-08-27 14:10:42.176473412 +0200 CEST [manager] container [postdeploy-2719] (6128d63ffcff1013ac9e57c1) has stopped
```

### Mais

#### Un impact non négligeable du traitement de migration sur le temps de réponse

Bien que l'opération soit un succès, nous avons couru plusieurs risques :

- la migration des données (phase `UDPATE`) a sensiblement ralenti le temps de réponse global ;

> - la migration des données (phase `CREATE INDEX CONCURENTLY`), sur une période limitée à x minutes, a tellement impacté le temps de réponse que l'on peut considérer que le service était interrompu.

Reviewer 3: le create index concurrently (il manque un r d'ailleurs !), c'est pas de la migration de données? J'ai l'impression que c'est un mauvais copier coller

Author: Effectivement, ce n'est pas de la migration de données => corrigé

Grâce à des circonstances favorables, l'impact a été réduit :

- la migration a eu lieu en été, période de faible fréquentation de la plateforme ;
- l'interruption de service a été de faible durée et a eu lieu is à un horaire de faible fréquentation (nuit).

Nous avions envisagé (et testé) des mesures de mitigation du risque, notamment :

- la possibilité d'arrêter/reprendre le traitement de migration ;
- la capacité de réduire l'impact du traitement (pauses périodiques, réduction de la taille des batchs).

> Malgré cela, il reste un point faible dans cette approche: l'interruption de service.

Reviewer 3: oui mais quand même. On n'a jamais promis du vrai zero downtime. Conclure sur cette phrase pour cette sous-section c'est un peu tristoune. L'interruption de service a été réduite à son maximum de nos capacité, là est la réussite ! (et aussi je souligne que ça a tellement été bien fait que j'ai pas souvenir d'une pétouille en prod où il fallait d'urgence rollback ou je sais pas quoi). C'était maîtrisé !

Author: je fais référence à l'épisode dans la nuit où la BDD répondait presque plus (il n'y a que @jonathanperret qui l'ait vu). C'est pas dit qu'on ait autant de chance la prochaine fois. je vais récupérer la durée d'indispo qu'on ait une idée

Author: mais peut être que ce n'est pas clair, comment le rendre plus clair ?

Author: J'ai mis les chiffres (et une image): c'est moins effrayant que ça sonnait, tu as eu raison d'insister !

Reviewer 2: @laura-bergoens une remarque qui n'a rien à voir avec l'article mais bon puisqu'on est là : on n'a pas "promis" au départ qu'on arriverait à du _zero-downtime_ certes, mais d'un autre côté on a fait le choix de ne pas annoncer de période d'indisponibilité aux utilisateurs. Du coup, ça crée quand même un engagement à… ne pas créer d'indisponibilité 😉

### Diminuer les impacts possibles

Le contexte de faible fréquentation de la plateforme ne peut pas être pris comme un pré-requis. En particulier, l'existence d'un verrou exclusif sur un période de plusieurs minutes est problématique.

Nous pensons à :

- modifier notre approche en passant d'une colonne temporaire à une table temporaire ;
- rendre le traitement de migration paramétrable à chaud, en passant de variables d'environnement à une table de paramétrage.

### Généraliser la solution

La table qui a été migrée n'était pas référencée par d'autres tables.
Cela était implicite, car elle était la plus volumineuse.

Cependant, comme le rappelle Gerald Weinberg avec "Rudy's Rutabaga rule"
> Once you eliminate your number one problem, number two gets a promotion.

Notre problème n°2 est une table d'une volumétrie inférieure à celle migrée, mais d'une volumétrie suffisante pour être concernée. Et cette table est elle-même référencée par la table migrée.

La situation est similaire, mais seule l'implémentation nous dira si cela est fondé:

- deux colonnes à mettre à jour au lieu d'une ;

> - une contrainte de type FK au lieu d'une contrainte de type PK.

Reviewer 3: ```suggestion
- une contrainte de type FK au lieu d'une contrainte de type PK.
```
pas au lieu mais en plus ! car il y aura la PK de la table answers quand même

Author: Juste ! Corrigé

---
layout: post
title:  "Passer une PK de INT en BIGINIT"
date:   2021-08-30 10:10:42 +0200
categories: bdd
excerpt: "Changer le type d'une clef primaire non référencée d'une table volumineuse de INTEGER à BIG INTEGER en minimisant l'indisponibilité"
authors: pierre_top
---

## TL;DR

Dans quelques mois, la table la plus volumineuse (nombre de tuples) aurait épuisé les valeurs disponibles pour la clef primaire. Pour éviter [cette situation](https://twitter.com/dhh/status/1060565296048562177?lang=fr), nous souhaitions changer de type en minimisant l'impact utilisateur, en visant des migrations de BDD "zero-downtime".

Nous avons réussi et partageons avec vous nos apprentissages. Pour les impatients, la solution est [ici](##implementation).

Ceci dit, cela a eu lieu dans des circonstances particulières :

- table non référencée par d'autres tables ;

> - traffic réduit (vacances).

Reviewer 3: ```suggestion
- trafic réduit (vacances).
```
traffic c'est la version cup of tea 😄

Author: Corrigé

> Ce chantier continuera sur une autre PK, sans ces deux circonstances, et nous partagerons également les résultats dans un autre post.

Reviewer 3: ```suggestion
Ce chantier continuera sur une autre PK, sans ces deux circonstances, et nous partagerons également les résultats dans un autre post.
```
continuer me paraît maladroit, pour KE en l'occurence le chantier est terminé.
On peut dire que l'opération sera à renouveler sur une autre table volumineuse

Author: C'est vrai que de l'extérieur, le contexte "Tech-days chantier gestion BDD" manque

Je vais aussi corriger ligne 213
>Ces actions font partie d'un chantier qui a dans son périmètre d'autres tables.

Author: Corrigé

Il est bien sûr possible de modifier le type de cette propriété avec l'instruction native `ALTER COLUMN TYPE`, mais cela a un coût proportionnel au remplissage de la table. Une exécution de cette instruction sur les données de production, sur un environnement moins puissant, a duré 8 heures.

Nous souhaitions trouver une alternative, à savoir effectuer ce changement sans que les utilisateurs s'en rendent compte:

- pas interruption de service, même sous la forme d'une fenêtre de maintenance planifiée ;
- pas de dégradation sensible des temps de réponse pendant les opérations.

## Recherche de solution

Ce [thread Stack Overflow](https://stackoverflow.com/questions/54795701/migrating-int-to-bigint-in-postgressql-without-any-downtime/54796046) est un excellent point d'entrée sur le sujet.

Il existe deux solutions :

- soit utiliser une colonne temporaire ;
- soit utiliser une base de données temporaire, alimentée par une réplication logique.

> Nous avons choisi la colonne temporaire, car cette solution peut être mise en ouvre immédiatement dans notre environnement.

Reviewer 3: typo oeuvre

Author: Corrigé

En effet, nous utilisons les services d'un [PaaS](https://scalingo.com/) et sommes heureux de ne pas avoir à installer et maintenir nos bases de données. Le service spécifique de la réplication logique n'est pour l'instant pas offert.

> ### Etape 1 : convertir les données en tâche de fond

Reviewer 3: suggestion : on s'embête à mettre les accents sur le É ou on s'en fiche ? Tout me va

Reviewer 2: ON MET COMPLÈTEMENT LES ACCENTS SUR LES CAPITALES PARCE QU'ON N'EST PAS DES BARBARES

Author: Corrigé

La [solution 1](https://tech.coffeemeetsbagel.com/schema-migration-from-int-to-bigint-on-a-massive-table-in-postgres-aabb835c3b84) effectue en tâche de fond la conversion des données implicite dans `ALTER TABLE foo COLUMN idBigInteger BIG INTEGER`.

La [documentation](https://www.postgresql.org/docs/current/sql-altertable.html#notes) mentionne
> Changing the type of an existing column will require the entire table and its indexes to be rewritten.
> As an exception, when changing the type of an existing column and the old type is either binary coercible to the new type a table rewrite is not needed; but any indexes on the affected columns must still be rebuilt.
> Table and/or index rebuilds may take a significant amount of time for a large table; and will temporarily require as much as double the disk space.

> Il pourrait sembler qu'aucune conversion n'est nécessaire de `INTEGER` à BIGINTEGER`, mais ce n'est pas le cas.

Reviewer 3: manque un tick au début de bigint

Author: Corrigé

```sql SELECT 1::integer::bigint```

Comme l'opération native pose un verrou exclusif sur la table, il faut attendre sa réécriture complète pour y accéder.
> An ACCESS EXCLUSIVE lock is held unless explicitly noted.

### Etape 2: substituer les colonnes et créer la clef primaire

> Une fois la migration achevée, une fenêtre de maintenance est requise (1h dans son cas) pour :

Reviewer 3: son cas ? c'est qui "son" ?

Author: Corrigé

Le principe est le suivant :

- la contrainte de clef primaire comporte une contrainte NOT NULL et une contrainte UNIQUE ;
- c'est la validation de ces contraintes qui prend le plus de temps dans la fenêtre de maintenance ;
- ces validations peuvent être effectuées en amont.

Pour valider les contraintes en amont :

- ajouter une contrainte NOT NULL dès la création de la colonne ;
- créer un index UNIQUE de manière concurrente [`CREATE INDEX CONCURRENTLY`](https://www.postgresql.org/docs/current/sql-createindex.html).

Lors de la création de la PK, il suffit de préciser l'index UNIQUE (PostgreSQL détecte la contrainte NOT NULL).
`ALTER TABLE "foo" ADD CONSTRAINT "foo_pkey" PRIMARY KEY USING INDEX "foo_id_unique`

## Implémentation

La solution est présentée en pseudo SQL/NodeJS sur une table imaginaire, avec les liens vers les commits réels.

On part de là

``` sql
CREATE TABLE foo (ID INTEGER PRIMARY KEY);
INSERT INTO foo generate_series(1, 1e6);
```

Voilà la solution native, avec accès exclusif sur la table pendant 8 heures.

```sql

> ALTER TABLE foo COLUMN idBigInteger BIG INTEGER;

Reviewer 3: ```suggestion
ALTER TABLE foo COLUMN "idBigInteger" BIG INTEGER;
```

Author: Corrigé partout avec camelCase et ""

> ALTER TABLE foo COLUMN idBigInteger BIG INTEGER;

Reviewer 3: ou
```suggestion
ALTER TABLE foo COLUMN id_big_integer BIG INTEGER;
```

Author: Corrigé partout avec camelCase et ""

> ALTER TABLE foo COLUMN idBigInteger BIG INTEGER;

Reviewer 3: car j'ai l'impression que le but c'est d'avoir un truc en ready to paste

Author: Pour cela, il y a les références au commits
Si je mets la version SQL réelle, rien que la création du trigger prend 20 lignes

```

> Tout commence avec une migration, lancée lors du déploiement de la release.

Reviewer 3: manque ici de dire que tu passes à la solution no downtime, je trouve que la séparation avec la ligne juste au dessus est pas très claire (on dirait une suite d'instructions alors qu'en fait là tu présentes la deuxième solution)

Author: Tu as raison, j'ai enlevé la ligne

<https://github.com/1024pix/pix/pull/3357/files#diff-1eecfd94150cca766655d3786c7dc12e3b6bc8fa60be3a0c0740424ab02c157a>

```sql

> ALTER TABLE foo ADD COLUMN idBigInteger NOT NULL DEFAULT -1;

Reviewer 3: idem pour la casse de la colonne que la remarque du dessus

Reviewer 3: et du coup je vais pas mettre un commentaire à chaque requête, you got the gist

Author: Corrigé partout avec camelCase et ""

---
layout: post
title:  "Passer une PK de INT en BIGINT sur PostgreSQL"
date:   2021-08-30 10:10:42 +0200
categories: bdd
excerpt: "Sur une table volumineuse, en minimisant l'indisponibilité"
cover:
  image: "change-pk-unreferenced-bigint.png"
  source_title: https://richmondlockandkey.com/master-key-systems/
  source_url: https://richmondlockandkey.com/wp-content/uploads/2020/04/unnamed-e1587059843528.png.webp
authors: pierre_top
---

## TL;DR

> Dans quelques mois, la table la plus volumineuse (nombre de tuples) aurait épuisé les valeurs disponibles pour la clef primaire. Pour éviter [cette situation](https://twitter.com/dhh/status/1060565296048562177?lang=fr), nous souhaitions changer de type en minimisant l'impact utilisateur, en visant des migrations de BDD "zero-downtime".

Reviewer 2: Suggestion : "nombre de lignes" ou "nombre d'enregistrements" plutôt que "tuples" qui est un anglicisme pas très heureux.

Reviewer 2: Pour **sa** clé primaire.

Author: Tiens, en français ça se dit [n-uplet](https://fr.wikipedia.org/wiki/N-uplet)
=> Corrigé

Reviewer 2: Oui ça se dit `n-uplet`, et une fois qu'on se débarrasse de l'anglicisme on se rend mieux compte qu'on est en train d'utiliser un terme que personne ne connaît pour désigner un concept basique. Note que la documentation PG alterne entre `rows` et `tuples` suivant les contextes — j'ai l'impression (mais ce n'est peut-être pas correct à 100%) qu'ils n'utilisent `tuples` que lorsqu'il est question des détails de stockage. Bref…

Nous avons réussi et partageons avec vous nos apprentissages. Pour les impatients, la solution est [ici](#implémentation).

Ceci dit, cela a eu lieu dans des circonstances particulières :

- table non référencée par d'autres tables ;
- trafic réduit (vacances).

Nous avons prévu de migrer d'autres tables, référencées, en charge nominale. Nous partagerons les résultats dans un autre article.

## Motivation

En un an, depuis septembre 2020, la fréquentation de Pix a augmenté [de manière spectaculaire](https://pix.fr/statistiques). La charge a été bien absorbée par l'infrastructure, mais nous nous sommes rendu compte d'une autre conséquence: une table allait épuiser l'identifiant utilisé par sa clef primaire.

> Il est bien sûr possible de modifier le type de cette propriété avec l'instruction native `ALTER COLUMN TYPE`, mais cela a un coût proportionnel au remplissage de la table. Une exécution de cette instruction sur les données de production, sur un environnement moins puissant, a duré 8 heures.

Reviewer 2: Suggestion: L'instruction `ALTER TABLE … ALTER COLUMN … SET TYPE`, c'est plus long mais bon `ALTER COLUMN TYPE` ça correspond à rien je trouve.

Author: Corrigé

> Nous souhaitions trouver une alternative, à savoir effectuer ce changement sans que les utilisateurs s'en rendent compte:

Reviewer 2: **ne** s'en rendent compte

Author: Corrigé

- pas d'interruption de service, même sous la forme d'une fenêtre de maintenance planifiée ;
- pas de dégradation sensible des temps de réponse pendant les opérations.

## Recherche de solution

Ce [thread Stack Overflow](https://stackoverflow.com/questions/54795701/migrating-int-to-bigint-in-postgressql-without-any-downtime/54796046) est un excellent point d'entrée sur le sujet.

> Il existe deux solutions :

Reviewer 2: "Deux solutions sont proposées dans ce thread" plutôt ?

Author: Corrigé

- soit utiliser une colonne temporaire ;
- soit utiliser une base de données temporaire, alimentée par une réplication logique.

Nous avons choisi la colonne temporaire, car cette solution peut être mise en œuvre immédiatement dans notre environnement.

En effet, nous utilisons les services d'un [PaaS](https://scalingo.com/), Scalingo, et sommes heureux de ne pas avoir à installer et maintenir nos bases de données. Le service spécifique de la réplication logique n'est pour l'instant pas offert.

Examinons ensemble les étapes de cette [solution](https://tech.coffeemeetsbagel.com/schema-migration-from-int-to-bigint-on-a-massive-table-in-postgres-aabb835c3b84) avec colonne temporaire.

### Etape 1 : convertir les données en tâche de fond

> La commande `ALTER TABLE foo COLUMN idBigInteger BIG INTEGER` effectue implicitement une conversion des données (cast) de INTEGER à BIGINT.

Reviewer 2: Comme plus haut, si on évoque une commande SQL j'aimerais autant qu'elle soit à peu près correcte, i.e. ici `ALTER TABLE foo ALTER COLUMN id SET TYPE bigint` (le type `BIG INTEGER` n'existe pas, notamment).

Author: Corrigé

La [documentation](https://www.postgresql.org/docs/current/sql-altertable.html#notes) mentionne en effet :
> Changing the type of an existing column will require the entire table and its indexes to be rewritten.
> As an exception, when changing the type of an existing column and the old type is either binary coercible to the new type a table rewrite is not needed; but any indexes on the affected columns must still be rebuilt.
> Table and/or index rebuilds may take a significant amount of time for a large table; and will temporarily require as much as double the disk space.

Comme l'opération native pose un verrou exclusif sur la table, il faut attendre sa réécriture complète pour y accéder.

> > An ACCESS EXCLUSIVE lock is held unless explicitly noted.

Reviewer 2: C'est intéressant de mentionner la solution "naïve" et pourquoi elle ne répond pas à nos critères, mais c'est étonnant de trouver ça dans une phase où on a déjà "choisi la colonne temporaire" ?
En fait ce paragraphe est redondant avec celui qui commence par "Il est bien sûr possible de modifier le type de cette propriété avec l'instruction native…" plus haut non ?

Author: Pas tout à fait. Je cherche à dire qu'il y a plusieurs tâches à effectuer:
- réécriture de la table
- réécriture des index
- recréation des contraintes.

Cela s'applique à la solution native, à celle de l'article original (coffee & bagels), et à la nôtre. 

Je veux :
- expliquer telle solution est plus efficace
- pointer les différences.

Notamment, je veux séparer ce qui était déjà fait et ce que l'on a ajouté  c'est à dire les index et contraintes avant la fenêtre de maintenance.

Ceci dit, ta remarque met en avant le fait que ce n'est pas explicité. Je garde donc ce paragraphe mais

Reviewer 2: "Ce qui était déjà fait" ? Je suis perdu parce qu'on n'utilise rien de la solution `ALTER TABLE … SET TYPE`. Mais bon, si ça te paraît compréhensible, all I can say is 🤷‍♂️

L'alternative à la solution native est d'effectuer la conversion des données en tâche de fond :

- dans une colonne de type BIGINT, non exposée à l'API ;
- pour les nouveaux tuples, à l'aide d'un trigger ;

> - pour les tuples existants, par un traitement batch à l'initiative de l'API.

Reviewer 2: tuples -> lignes

Reviewer 2: Que veut dire "à l'initiative de l'API" ici ? Si je ne connaissais pas le contexte, en lisant ça je penserais à une solution où du code dans l'API va mettre à jour des lignes sous certaines conditions… alors que justement, on a fait toute la migration sans toucher au code de l'API !

Author: Ab-so-lu-ment d'accord. Je me suis posé la même question d'interprétation. 

Je cherchais à dire que le traitement est fait en partie hors de la BDD (par opposition à la première version en SQL pur). Hop, je supprime et on en parle plus.

- la contrainte de clef primaire comporte une contrainte NOT NULL et une contrainte UNIQUE ;
- c'est la validation de ces contraintes qui prend le plus de temps dans la fenêtre de maintenance ;
- ces validations peuvent être effectuées en amont.

Pour valider les contraintes en amont :

- ajouter une contrainte NOT NULL dès la création de la colonne ;
- créer un index UNIQUE de manière concurrente [`CREATE INDEX CONCURRENTLY`](https://www.postgresql.org/docs/current/sql-createindex.html).

Il ne reste plus, comme dernière opération, qu'à créer la PK sur base de l'index UNIQUE.
`ALTER TABLE "foo" ADD CONSTRAINT "foo_pkey" PRIMARY KEY USING INDEX "foo_id_unique"`

PostgreSQL détecte lui-même la contrainte NOT NULL.

## Implémentation

Le code source étant trop verbeux pour figurer intégralement dans cet article, la solution est présentée en pseudo SQL/NodeJS sur une table imaginaire, avec un lien vers le commit de notre repository pour les détails.

A l'état initial, la table contient 700 millions d'enregistrements.
[fd7273f9](https://github.com/1024pix/pix/commit/fd7273f9e47648e4615a8b4af13bd81b1c8c2928#diff-754627687e8d1bd3c4d8f60cf5ed3640a9c104acdccbc20407c16479ad63055d)

``` sql

> CREATE TABLE foo (ID INTEGER PRIMARY KEY);
> INSERT INTO foo generate_series(1, 700 000 000);

Reviewer 2: ```suggestion
CREATE TABLE foo (id integer PRIMARY KEY);
INSERT INTO foo SELECT * FROM generate_series(1, 700000000);
```
Ça ne coûte pas vraiment plus cher de respecter la syntaxe…

Author: Voir https://github.com/1024pix/pix-engineering/pull/19#discussion_r701637398

```

La première étape est l'ajout d'une colonne temporaire, avec une migration de BDD, inclue dans le déploiement de la release.
[eecfd941](https://github.com/1024pix/pix/pull/3357/files#diff-1eecfd94150cca766655d3786c7dc12e3b6bc8fa60be3a0c0740424ab02c157a)

```sql

> ALTER TABLE foo ADD COLUMN "idBigInteger" NOT NULL DEFAULT -1;
> CREATE TRIGGER WHEN INSERT INTO foo FOR EACH ROW :NEW."idBigInteger" = :new.id;

Reviewer 2: ```suggestion
ALTER TABLE foo ADD COLUMN "idBigInteger" bigint NOT NULL DEFAULT -1;
CREATE TRIGGER trg_bigint BEFORE INSERT ON foo FOR EACH ROW EXECUTE FUNCTION copy_int_id_to_bigint_id();
```
On peut donner le code de `copy_int_id_to_bigint_id` (dont la déclaration est certes un peu verbeuse), ou pas ; mais je ne suis pas vraiment convaincu par le pseudo-SQL que je trouve à la fois trop abscons pour du pseudo-code et perturbant pour qui connaît la syntaxe de PostgreSQL (première réaction de mon cerveau : "Tiens, c'est `CREATE TRIGGER WHEN` la syntaxe ? Il me semblait que c'était plutôt `ON` 🤔 Tiens je vais aller lire la doc…" et à ce stade je ne suis plus en train de lire l'article 😅).

Author: L'argument "Tiens je vais aller lire la doc" me semble un peu exagéré, mais cela rejoint des remarque de @laura-bergoens . J'ai donc remis le code du repo, sauf quand il y avait du knex, auquel cas j'ai traduit en SQL.

```

Ensuite, le script de migration inclus dans la PR est lancé manuellement (durée: 48h pour 700 millions de lignes).
[d89cb206](https://github.com/1024pix/pix/pull/3357/files#diff-d89cb20637cb9b836bce4902b46db958e0209c2a0de96e7dc81a0c43015680b4)

> ```javascript
> const maxid = await client.query('SELECT MAX(id) INTO maxId FROM foo');
> for (const i = 0; i <= maxId; i++){
>   await client.query(`UPDATE foo SET "idBigInteger" = id WHERE id BETWEEN i AND (i + ${chunkSize} -1)`);
> }
> await client.query('CREATE UNIQUE INDEX CONCURRENTLY "indexidBigInteger"');
> ```

Reviewer 2: ```suggestion
const chunkSize = 10000000;
const {rows: [{maxId}]} = await client.query('SELECT MAX(id) "maxId" FROM foo');
for (let startId = 0; startId <= maxId; startId += chunkSize) {
  await client.query(`UPDATE foo SET "idBigInteger" = id WHERE id BETWEEN ${startId} AND ${startId + chunkSize - 1}`);
}
await client.query('CREATE UNIQUE INDEX CONCURRENTLY "idx_idBigInteger" ON foo("idBigInteger")');
```

Même opinion sur le pseudo-JS que sur le pseudo-SQL. Je ne pense pas que glisser des erreurs comme `SELECT MAX(id) INTO` ou `for(const i` aide à la lecture. On peut proposer du code simplifié qui reste correct.

Author: Yep, voir https://github.com/1024pix/pix-engineering/pull/19#discussion_r701637398

Pour finir, une dernière migration, celle qui expose la colonne BIGINT
[9985cafe](https://github.com/1024pix/pix/pull/3364/files#diff-9985cafe7c9915c64927036180071c804f75aad26a66686efd1758d242b819d1)

```sql
BEGIN TRANSACTION;
LOCK foo WITH ACCESS EXCLUSIVE;
ALTER SEQUENCE "seqPkFoo" TYPE BIG INTEGER;
ALTER TABLE foo ALTER COLUM "idBigInteger" SET DEFAULT "seqPkFoo".nextval();
ALTER TABLE foo DROP COLUMN id;
ALTER TABLE foo RENAME COLUMN "idBigInteger" TO id;
ALTER TABLE foo ADD PRIMARY KEY id USING "indexidBigInteger";

> COMMIT TRANSACTION;

Reviewer 2: ```suggestion
BEGIN TRANSACTION;
LOCK TABLE foo IN ACCESS EXCLUSIVE MODE;
ALTER SEQUENCE "seqPkFoo" AS bigint;
ALTER TABLE foo ALTER COLUMN "idBigInteger" SET DEFAULT nextval("seqPkFoo");
ALTER TABLE foo DROP COLUMN id;
ALTER TABLE foo RENAME COLUMN "idBigInteger" TO id;
ALTER TABLE foo ADD CONSTRAINT "foo_pkey" PRIMARY KEY USING INDEX "idx_idBigInteger";
COMMIT TRANSACTION;
```

Author: Yep, voir https://github.com/1024pix/pix-engineering/pull/19#discussion_r701637398

## Exécution en production et perspectives

### Le plan se déroule sans accroc

La migration préalable s'est déroulée sans erreur.

La migration des données proprement dite a pris environ 24h.

L'exposition de la nouvelle colonne, lors du déploiement, a pris un temps négligeable pour les utilisateurs finaux (< 1 seconde).

```shell
2021-08-27 14:10:40.368443479 +0200 CEST [postdeploy-2719] > pix-api@3.92.0 postdeploy /app
2021-08-27 14:10:40.368393446 +0200 CEST [postdeploy-2719]
2021-08-27 14:10:41.566685265 +0200 CEST [postdeploy-2719] Batch 188 run: 1 migrations
2021-08-27 14:10:42.176473412 +0200 CEST [manager] container [postdeploy-2719] (6128d63ffcff1013ac9e57c1) has stopped
```

### Mais

#### Un impact non négligeable du traitement de migration sur le temps de réponse

Bien que l'opération soit un succès, lors des étapes préparatoires :

* le temps de réponse de l'API a augmenté d'un facteur 10, en passant d'une moyenne de 30 ms à 300 ms ;
* lors de l'étape de migration des données, ont eu lieu 4 pics à 1 seconde, d'une demi-heure chacun ;
* lors de l'étape de validation des contraintes, a eu lieu 1 pic à 1 seconde, d'une heure et demie.

> ![image info](/assets/images/posts/change-pk-unreferenced-bigint/average-response-time.png)

Reviewer 2: Regarder le temps moyen c'est plutôt optimiste comme vision, il me semble que la norme quand on s'intéresse aux temps de réponse est plutôt de regarder quelque chose comme le 95ème centile. Ça change pas mal les chiffres.

Author: Corrigé :fearful:

En effet :

- la solution native réécrit la table en une seule fois ;
- notre solution la réécrit en plusieurs fois.

Cela peut avoir deux effets :

- occupation de disque plus importante ;
- impact sur les temps de réponse.

Nous monitorons actuellement les temps de réponse pour déterminer s'il y a un impact.

## Perspectives

Nous avons prévu de changer le type d'identifiant sur d'autres tables.
Nous ne pourrons pas appliquer exactement la même approche à l'avenir, pour deux raisons qui seront détaillées. Aussi, nous allons continuer à y travailler et cela fera l'objet d'un nouvel article.

### Diminuer les impacts possibles

Le contexte de faible fréquentation de la plateforme ne peut pas être pris comme un pré-requis. En particulier, l'existence d'un verrou exclusif sur un période de plusieurs minutes est problématique.

Nous pensons à :

- modifier notre approche en passant d'une colonne temporaire à une table temporaire ;
- rendre le traitement de migration paramétrable à chaud, en passant de variables d'environnement à une table de paramétrage.

### Généraliser la solution

La table qui a été migrée n'était pas référencée par d'autres tables.

> Cela était implicite, car elle était la plus volumineuse.

Reviewer 2: Hmm… tu veux dire que si une table est plus volumineuse (en nombre de lignes) que toutes les autres, elle ne peut pas être référencée par d'autres tables ? Une sorte de [Principe des tiroirs](https://fr.wikipedia.org/wiki/Principe_des_tiroirs) à l'envers ? Ça serait intéressant à défendre 😉 mais je ne suis pas sûr que ce soit clair pour le lecteur ici.

Author: Il y a effectivement (diable, ce Jonathan voit clair dans mon jeu !) des hypothèses implicites:
- la FK est NOT NULL
- la majorité des valeurs est référencée au moins une fois.

Ca fait beaucoup d'implicite, je l'ai enlevé.

Cependant, comme le rappelle Gerald Weinberg avec "Rudy's Rutabaga rule"
> Once you eliminate your number one problem, number two gets a promotion.

Notre problème n°2 est une table d'une volumétrie inférieure à celle migrée, mais d'une volumétrie suffisante pour être concernée. Et cette table est elle-même référencée par la table migrée.

La situation est similaire, mais seule l'implémentation nous dira si cela est fondé:

- deux colonnes à mettre à jour au lieu d'une ;
- une contrainte de type FK en plus de la contrainte de type PK.

## Des suggestions d'amélioration ?

Afin de rester sur un format court, cet article :

- prend comme pré-requis le MVCC de PostgreSQL, et ses conséquences lors d'ajout et de suppression de colonne (bloat) ;
- ne précise pas la raison des détails d'implémentation, notamment du script de migration des données.

> Si vous avez des questions, contactez-nous à service@pix.fr pour que nous puissions améliorer cet article !

Reviewer 2: service@pix.fr est réservé aux comptes créés sur des services tiers, pas aux communications humaines.

dev@pix.fr semble fonctionner y compris depuis une adresse extérieure à Pix.

Author: Corrigé

Reviewer 4: ℹ️ il faut utiliser engineering@pix.fr

Author: J'ai switché en demandant d'ouvrir une issue sur le repo du blog, ça me semble mieux, non ?

---
layout: post

> title:  "Passer une PK de INT en BIGINT en zéro downtime"

Reviewer 5: Petite question, est ce que ce titre apparaît dans un h1 sur la page générée ?

Author: Ouaip, pour être précis
```
        <h1 class="article-header__title">Passer une PK de INT en BIGINT en zéro downtime</h1>
        <div class="article-header__subheader">
          <div class="article-header__authors">
            
            
            <div class="article-header__author">
              <img class="article-header__author-picture" src="/assets/images/authors/pierre_top.png"
                   alt="Pierre Top">
              <span class="article-header__author-name">Pierre Top</span>
            </div>
            
          </div>
          <div class="article-header__date"><time class="article-header__data-time" datetime="2021-08-30T10:10:42+02:00">
              30 Aug 2021
            </time>
          </div>
        </div>
      </div>
    </header>
```

Il est bien sûr possible de modifier le type de cette propriété avec l'instruction native `ALTER TABLE foo ALTER COLUMN id SET DATA TYPE BIGINT`, mais cela a un coût proportionnel au remplissage de la table. Une exécution de cette instruction sur les données de production, sur un environnement moins puissant, a duré 8 heures.

Nous souhaitions trouver une alternative, à savoir effectuer ce changement sans que les utilisateurs ne s'en rendent compte :

- pas d'interruption de service, même sous la forme d'une fenêtre de maintenance planifiée ;
- pas de dégradation sensible des temps de réponse pendant les opérations.

## Recherche de solution

Ce [thread Stack Overflow](https://stackoverflow.com/questions/54795701/migrating-int-to-bigint-in-postgressql-without-any-downtime/54796046) est un excellent point d'entrée sur le sujet.

Deux solutions y sont proposées :

- soit utiliser une colonne temporaire ;
- soit utiliser une base de données temporaire, alimentée par une réplication logique.

Nous avons choisi la colonne temporaire, car cette solution peut être mise en œuvre immédiatement dans notre environnement.

> En effet, nous utilisons les services d'un [PaaS](https://scalingo.com/), Scalingo, et sommes heureux de ne pas avoir à installer et maintenir nos bases de données. Le service spécifique de la réplication logique n'est pour l'instant pas offert.

Reviewer 6: **suggestion** : décaler le lien de Scalingo sur le mot Scalingo et non sur PaaS. Sur le mot PaaS si il y a un lien je m'attends à ce que cela soit une définition.

Author: Corrigé

Examinons pour commencer la [solution d'origine](https://tech.coffeemeetsbagel.com/schema-migration-from-int-to-bigint-on-a-massive-table-in-postgres-aabb835c3b84), celle avec colonne temporaire, puis voyons les modifications que nous y avons apportées.

### Étape 1 : convertir les données en tâche de fond

> La commande `ALTER TABLE foo ALTER COLUMN idBigInteger SET DATA TYPE BIGINT` effectue (entre autres) une conversion des données (cast) de INTEGER à BIGINT.

Reviewer 2: ⚠️  Si on utilisait `ALTER TABLE…SET TYPE bigint` ce serait sur la colonne existante `id`, pas sur une colonne temporaire `idBigInteger` !

Author: Corrigé

- c'est la validation de ces contraintes qui prend le plus de temps dans la fenêtre de maintenance ;
- ces validations peuvent être effectuées en amont.

Pour valider les contraintes en amont :

- ajouter une contrainte NOT NULL dès la création de la colonne ;
- créer un index UNIQUE de manière concurrente [`CREATE INDEX CONCURRENTLY`](https://www.postgresql.org/docs/current/sql-createindex.html).

Il ne reste plus, comme dernière opération, qu'à créer la PK sur base de l'index UNIQUE.
`ALTER TABLE "foo" ADD CONSTRAINT "foo_pkey" PRIMARY KEY USING INDEX "foo_id_unique"`

PostgreSQL détecte lui-même la contrainte NOT NULL.

## Implémentation

Le code présenté a été modifié pour les besoins de l'article, mais un lien est fourni vers chaque commit pour les détails.

A l'état initial, la table contient 700 millions d'enregistrements.
[fd7273f9](https://github.com/1024pix/pix/commit/fd7273f9e47648e4615a8b4af13bd81b1c8c2928#diff-754627687e8d1bd3c4d8f60cf5ed3640a9c104acdccbc20407c16479ad63055d)

``` sql
CREATE TABLE "knowledge-elements"(
    id             SERIAL PRIMARY KEY,
    source         varchar(255),
    (...)
);
```

> La première étape est l'ajout d'une colonne temporaire, avec une migration de BDD, inclue dans le déploiement de la release.

Reviewer 5: Si je ne dis pas de bêtises c'est "incluse" (quelqu'un peut confirmer 😂  ?).

```suggestion
La première étape est l'ajout d'une colonne temporaire, avec une migration de BDD, incluse dans le déploiement de la release.
```

Author: [elle](https://www.lalanguefrancaise.com/orthographe/inclue-ou-incluse) confirme

ALTER TABLE "knowledge-elements" ALTER COLUMN "id" DROP DEFAULT';
ALTER TABLE "knowledge-elements" DROP CONSTRAINT "knowledge-elements_pkey";
ALTER TABLE "knowledge-elements" ALTER COLUMN "id" DROP NOT NULL;
ALTER TABLE "knowledge-elements" ADD CONSTRAINT "knowledge-elements_pkey" PRIMARY KEY USING INDEX "knowledge-elements_bigintId_index";
ALTER TABLE "knowledge-elements" RENAME COLUMN "id" TO "intId";
ALTER TABLE "knowledge-elements" RENAME COLUMN "bigintId" TO "id";
COMMIT TRANSACTION;
```

## Validation de la solution

Cette hypothèse de solution a été mise progressivement en contact avec la réalité, pour garder un feedback rapide :

- exécution sur une table de test, ne comportant qu'une seule colonne, sur une volumétrie comparable à la production ;
- même test, en soumettant la BDD à des requêtes concurrentes de lecture/écriture sur la table pour estimer l'impact sur les temps de réponse ;
- exécution sur une copie de la table de production ;
- même test, en soumettant la BDD à un échantillon des requêtes de production sur la table pour estimer l'impact sur les temps de réponse ;
- intégration du code dans le processus de migration de BDD (knex) ;

> - déploiement du code, via le processus de migration, sur une instance de reporting de production.

Reviewer 5: Merci j'ai appris qu'il faut un espace avant le point virgule 😮  !

Author: C'est mon spellchecker qui m'a appris :smile: 
J'ai quand même vérifié vu que je en savais pas s'il confondait anglais et français
https://www.lalanguefrancaise.com/articles/espace-insecable

Reviewer 2: Oui alors si on pinaille il faut _une_ espace bien sûr 😉
